python/lsst/ts/salkafka/component_producer_set.py

```diff
@@ -116,17 +176,113 @@ def signal_handler(self):
+                    producer_set.append(
+                        loop.run_in_executor(
+                            pool,
+                            functools.partial(
+                                cls.create_process,
+                                args=args,
+                                queue_len=component_info.get(
+                                    "queue_len", salobj.topics.DEFAULT_QUEUE_LEN
+                                ),
+                                component=component_info["component"],
+                                add_ack=topic_set.get("add_ack", False),
+                                commands=topic_set.get("commands", []),
+                                events=topic_set.get("events", []),
+                                telemetry=topic_set.get("telemetry", []),
+                            ),
+                        )
+                    )
+
+                # In case one of the parallel producers fails, need to cancel
+                # all the others and exit. This will process any task that
+                # finished and then, proceed to cancel the remaining.
+                main_tasks = [wait_forever_task] + producer_set
+
+                for completed in asyncio.as_completed(main_tasks):
+                    try:
+                        await completed
+                    except asyncio.CancelledError:
+                        print("Terminating process.")
+                    except Exception:
+                        traceback.print_exc()
+                    finally:
+                        break
+
+                # If we get here, it means that it either received a term
+                # signal or one of the child processes failed.
+                # Now it need to send each remaining child process a TERM
+                # signal as well, before exiting.
+                # Gather information about remaining child processes and send
+                # signal.
+
+                # This is the parent process id.
+                parent = psutil.Process(os.getpid())
+
+                # Child process ids.
+                children = parent.children(recursive=True)
+
+                # Send SIGTERM regardless of what signal was received. This
+                # is the safest signal.
+                for process in children:
+                    print(f"Killing child process {process}.")
+                    process.send_signal(signal.SIGTERM)
+
+                # Wait for processes to terminate, skip any exception
+                print("Waiting for child processes to finish.")
+                await asyncio.gather(*producer_set, return_exceptions=True)
+                print("Done")
+
+    @classmethod
+    def create_process(
+        cls, args, queue_len, component, add_ack, commands, events, telemetry
+    ):
+        loop = asyncio.new_event_loop()
+
+        asyncio.set_event_loop(loop)
+
+        producer_set = ComponentProducerSet(
```

Comment:
This should be `producer_set = cls(`

Please look for `ComponentProducerSet` everywhere in this file and:
* In all classmethods replace it with `cls`
* In all staticmethods change that to classmethod and see previous item